Support for np.partition

rjenc29 opened the pull request:
<!--

Thanks for wanting to contribute to Numba :)

First, if you need some help or want to chat to the core developers, please
visit https://gitter.im/numba/numba for real time chat or post to the Numba
mailing list https://groups.google.com/a/continuum.io/forum/#!forum/numba-users.

Here's some guidelines to help the review process go smoothly.

0. Please write a description in this text box of the changes that are being
   made.

1. Please ensure that you have written units tests for the changes made/features
   added.

2. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

3. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]` and
   then remove the label when you'd like it to be reviewed.

4. Once review has taken place please do not add features or make changes out of
   the scope of those requested by the reviewer (doing this just add delays as
   already reviewed code ends up having to be re-reviewed/it is hard to tell
   what is new etc!). Further, please do not rebase your branch on master/force
   push/rewrite history, doing any of these causes the context of any comments
   made by reviewers to be lost. If conflicts occur against master they should
   be resolved by merging master into the branch used for making the pull
   request.

Many thanks in advance for your cooperation!

-->

stuartarchibald commented:
Thanks for opening this PR. The failures from 32bit linux coming from the Azure builds can be ignored at present whilst issues with the CI instructions are sorted out!

rjenc29 commented:
Sure thing – great to see the nascent Azure builds!

So, on to this PR….  

In the interests of full disclosure:

1. I struggled a bit to get my head around the behaviour of ``np.partition`` in the general case (multi-dimensional array containing some NaNs etc).
2. I haven’t attempted to replicate the ‘introselect’ algorithm; I just thought I’d wire in the existing select / partition algos which median and percentile use. 

Given (1) and (2), I decided to reimplement all of the (relevant) unit tests from numpy and add a few of my own – therefore, the tests are in general not comparing versus numpy output, but rather asserting certain properties of the output which must be true for a correctly partitioned array (of which there may be many for a given input array and partition index). 

The performance is marginally better or the same as numpy for ‘small’ problems but doesn’t scale particularly well and is outperformed by numpy for ‘not small’ problems.  I can provide more info / examples / a notebook if helpful.

From a review perspective, whenever you have some time:

1. The output will not in general match numpy except with regards to those properties which be true for a correctly partitioned array (in my understanding).  Is this acceptable?
2. The scaling / performance properties are not great – I can think of some areas which might contribute but some core dev advice would be appreciated.
3. Should we look to implement ``np.argpartition`` whilst we're at it? 
4. The usual stuff, please :)

seibert commented:
FYI: The Linux 32 failure on Azure can be fixed if you pull #3309 into this PR.

stuartarchibald commented:
@rjenc29 Thanks for this. To your points...
1. I think this is ok, unless the algorithms in NumPy are replicated exactly matching will be very hard. Further the guarantee over the output is `The ordering of the elements in the two partitions is undefined.`, so users shouldn't be relying on output order.
2. I think that two things are adversely impacting the performance, the first is the NaN checking, the second is the reuse of `_select` which doesn't seem to have the same level of efficiency as NumPy's equivalent. I'm of the view that whilst it'd be great to match or beat the performance of NumPy, first, let's make sure this is all working and there's an implementation to work from.
3. Could do, though I think let's get this in first as I expect any changes/improvements to this would be reflected in `argpartition`?

rjenc29 commented:
Thanks for the feedback.  I tried out a couple of options to reduce the cost of the NaN checking / number of pivots & I think you're on the money.

The last commit provides a speedup over my original of x2 to x5, depending on the nature of the inputs, and means there's a wider range of problem sizes for which the performance is the same or better than NumPy. 

It does however involve touching the `_partition` algorithm - in theory, I haven't changed its default behaviour (functional or non functional), but it's reached quite widely so there's attendant risk.  It also might look a bit weird.

It would look less weird & be less risky if I instead implemented a 'nan aware partition' but it might look a bit duplicitous as it's largely 'the same' except for the predicate which is used to determine whether a pivot is needed.

Anyway, if you have time to take a look, let me know what you reckon.

stuartarchibald commented:
@rjenc29 no problem, thanks for looking at this again, glad to hear that my guess at performance bottlenecks was correct and some speed up was achieved. I think changing the `_partition` alg is probably a necessity here, baking the NaN handling into the pivoting saves multiple array walking operations and multiple branches for result computation. To avoid the duplication/propagating kwargs factory functions may help as named algs can be trivially composed with them, have a look at: https://github.com/stuartarchibald/numba/commit/90acad4c9a5433a641375115759326eafd00be74 and see what you think, you are welcome to use/not use it!

rjenc29 commented:
Thanks for the factory functions suggestion - very neat.  Included in the last set of commits.

rjenc29 commented on numba/targets/arraymath.py:
```diff
@@ -967,6 +989,74 @@ def nanmedian_impl(arry):
+@overload(np.partition)
+def np_partition(a, kth):
+
+    if isinstance(kth, (types.Array, types.Sequence)):
```
Took out the branching and added support for Boolean (who knew)!

Put in a guard to reject 0D arrays and support array-like inputs (tuple, list).

rjenc29 commented on numba/targets/arraymath.py:
```diff
@@ -967,6 +989,74 @@ def nanmedian_impl(arry):
+            raise TypeError('Partition index must be integer')
+
+    def np_partition_impl(a, kth):
+        if len(a.flat) == 0:
```
Done

rjenc29 commented on numba/targets/arraymath.py:
```diff
@@ -967,6 +989,74 @@ def nanmedian_impl(arry):
+        if not isinstance(kth, types.Integer):
+            raise TypeError('Partition index must be integer')
+
+    def np_partition_impl(a, kth):
```
Agreed, now using _asarray.

Those two edge cases are pretty weird - I added them explicitly to check the behaviour is equivalently weird.

rjenc29 commented:
Added references to the original NumPy tests and a sanity check along the lines suggested which I have dropped in a few places.  

Added a bunch of additional tests - both positive (e.g. support for array-like inputs) and negative (reject 0D arrays).

The last CI failed but I suspect this is spurious.  I will commit something tomorrow morning to give CI a kick.

Cheers!

stuartarchibald commented:
Thanks @rjenc29 , I've given CI a shove. By the nature of the way it broke/got stuck, there may be a memory leak.

rjenc29 commented:
Great, thanks!  I will keep an eye out.

rjenc29 commented:
Added handling for another edge case (reject multi-dimensional kth) and some minor tweaks.

A few of the CI jobs timed out; those which did not have passed.  

I will hold off on any more changes at this time; if you have time to review, that would be great.

rjenc29 commented:
5 jobs timed out in the last CI run (4 on the previous).

Just a thought - there's a way of speeding up the case where kth elements are sequential - e.g. `np.partition(some_array, (1, 2, 3, 4))` - it's a bit fiddly, but it seems to work.  Anyway, I will park it for now rather than keep moving the review goal posts :)

seibert commented:
Aside: Our test duration has been creeping up on the Travis CI 50 minute maximum for a while, and it seems we've finally hit it.  Might be a good time to abandon it since the same tests finish in 27 minutes on Azure.

stuartarchibald commented:
Moving test duration discussion to https://github.com/numba/numba/issues/3391.

stuartarchibald commented on numba/tests/test_np_functions.py:
```diff
@@ -768,6 +772,322 @@ def test_triu_basic(self):
+
+    def test_partition_exception_out_of_range(self):
+        # inspired by the test of the same name in:
+        # https://github.com/numpy/numpy/blob/043a840/numpy/core/tests/test_multiarray.py
```
These links don't seem to resolve?! think the full hash might be needed?

stuartarchibald replied:
They now work! Mystery!!

stuartarchibald commented:
@sklam as this contains a variant of my patch https://github.com/stuartarchibald/numba/commit/90acad4c9a5433a641375115759326eafd00be74 please could you take a quick look. I'm of the view that this PR is good to go.

stuartarchibald commented:
@rjenc29 this has a conflict since merging https://github.com/numba/numba/pull/3337, please could you merge in/rebase on master? Thanks.

rjenc29 commented:
Done - will keep an eye on CI...

stuartarchibald commented:
@rjenc29 thanks for the fix up. @sklam confirmed "Looks good" over private channel (thanks!). This can be merged once CI passes.
